### PR TITLE
feat(shadcn): add SOCKS proxy support

### PR DESCRIPTION
**Pull request by nbouvrette**
## Summary

Adds **SOCKS4/SOCKS5** support to the registry HTTP stack via `ALL_PROXY`, closing the most commonly requested gap in proxy coverage.

> **Scope note:** this PR originally replaced `node-fetch` with `undici` and expanded `HTTPS_PROXY`/`HTTP_PROXY`/`NO_PROXY` handling. That portion shipped separately in #10905, and #11196 later added cross-origin redirect hardening. Per review feedback it has since been **rescoped again — PAC support has been removed and will land as a follow-up PR**, leaving SOCKS as the only new capability here. All redirect-security and error-enrichment logic from those PRs is preserved unchanged.

## Proxy capability matrix

| Capability | `main` today | After this PR |
|---|---|---|
| `HTTPS_PROXY` / `https_proxy` | ✅ | ✅ |
| `HTTP_PROXY` / `http_proxy` | ✅ | ✅ |
| `NO_PROXY` / `no_proxy` bypass | ✅ | ✅ |
| HTTPS proxy (CONNECT tunneling) | ✅ | ✅ |
| Cross-origin redirect header stripping | ✅ | ✅ |
| SOCKS4 / SOCKS5 (`ALL_PROXY=socks*://…`) | ❌ | ✅ |
| PAC files (`PAC_URL=…`) | ❌ | ➡️ follow-up PR |

Part of the gap tracked in #5952.

## Architecture

`proxy.ts` keeps its existing `fetchWithProxy` entry point and manual-redirect logic exactly as-is. The only change to that layer is the dispatcher it hands to `fetch`: the module-level `EnvHttpProxyAgent` constant is replaced by a `createProxyDispatcher(env)` factory.

The factory is a **strict superset** of the previous behavior — it gates HTTP/HTTPS on the identical four env vars and returns the identical `EnvHttpProxyAgent`, then adds one higher-priority branch ahead of it:

```
        ┌──────────────────────────────┐
        │  createProxyDispatcher(env)  │
        └──────────────┬───────────────┘
                       │ examines env vars
             ┌─────────┼──────────────┬─────────────┐
             ▼         ▼              ▼             ▼
        ALL_PROXY=   HTTP_PROXY /   no proxy
        socks*?      HTTPS_PROXY?   env set
             │              │             │
             ▼              ▼             ▼
        socks Agent    EnvHttp-       undefined
        (custom        ProxyAgent     (global
        undici Agent   (unchanged)    dispatcher
        using socks                   used)
        package)
```

`ALL_PROXY` only activates the SOCKS branch for `socks*://` schemes (the curl convention); `ALL_PROXY=http://…` is intentionally left to the explicit `HTTP_PROXY` variable rather than being silently absorbed.

### Code
- **`packages/shadcn/src/registry/proxy.ts`** — adds `createProxyDispatcher(env)`, the SOCKS `Agent` factory (a custom undici `Agent` whose `connect` is backed by the `socks` package), and the SOCKS URL parser. `fetchWithProxy` / `fetchOnce` / `getFailureReason` are untouched apart from the one-line dispatcher swap and one explanatory comment.

### Dependencies (`packages/shadcn/package.json`)
Adds `socks@^2.8.8` (~50 KB) and nothing else. `socks` was already present in the lockfile as a transitive dependency; promoting it to a direct dependency moves it `2.8.6 → 2.8.9`, which in turn updates its own `ip-address` dependency and drops two now-unused transitives. The lockfile diff is 32 lines and contains nothing else.

### Tests
- **`proxy.test.ts`** — 23 unit tests for `createProxyDispatcher` covering env-var detection (HTTP and SOCKS), priority ordering, and edge cases (empty values, malformed URLs, `ALL_PROXY` with a non-SOCKS scheme). The existing `fetchWithProxy` redirect tests are unchanged.
- **`proxy.integration.test.ts`** (new) — 10 integration tests against real local servers:
  - a CONNECT-tunneling HTTP proxy (verifies undici routes HTTPS via CONNECT)
  - a hand-rolled minimal SOCKS5 server (verifies SOCKS routing end to end, including that the SOCKS server observes a CONNECT to the correct origin host:port and that the HTTP proxy is not involved)

The CONNECT-tunnel test caught a real surprise during development: undici's `EnvHttpProxyAgent` always uses HTTP CONNECT even for `http://` targets, which a naive always-200 mock proxy would have hidden.

## Note for reviewers

`fetchOnce` must keep using the **global** `fetch` binding. `msw/node` patches `globalThis.fetch`, so importing `fetch` from `undici` there bypasses MSW's interceptor and breaks the registry tests — verified by a failed run during development. An inline comment documents this so future refactors don't regress it.

## Test plan

- [x] `pnpm typecheck` passes
- [x] `pnpm test` — all **1711 tests across 83 files** in `packages/shadcn` pass
- [x] New proxy tests (33) cover env-var detection and end-to-end routing through HTTP and SOCKS
- [x] Existing MSW-backed registry tests (`fetcher.test.ts`, `api.test.ts`, `resolver.test.ts`) still pass — confirms MSW interception works through the new dispatcher
- [x] Existing `fetchWithProxy` redirect tests from #11196 still pass unchanged
- [x] `pnpm install --frozen-lockfile` validates cleanly against the regenerated lockfile

### CODE REVIEW

**Comment by nbouvrette** [comment]
@shadcn sorry for the unexpected PR but I think this small change will reduce warnings and make the library more future proof.

**Comment by axieum** [comment]
This PR will resolve a lot of proxy-related issues I am having with `NO_PROXY=` support in shadcn/ui too.

Great stuff!

File: packages/shadcn/src/registry/fetcher.ts [hunk 1]
```diff
@@ -13,12 +13,14 @@ import {
...
+import { type Dispatcher, ProxyAgent } from "undici"
 import { z } from "zod"
 
-const agent = process.env.https_proxy
-  ? new HttpsProxyAgent(process.env.https_proxy)
+// MSW's Node adapter patches `globalThis.fetch`, so we keep using it here.
+// `dispatcher` is an undici-specific fetch option not declared on DOM's
+// RequestInit; we type it locally to pass it through without casting.
+const dispatcher: Dispatcher | undefined = process.env.https_proxy
+  ? new ProxyAgent(process.env.https_proxy)
```
Comment (pomeh) [inline packages/shadcn/src/registry/fetcher.ts]: In [my initial PR](https://github.com/shadcn-ui/ui/pull/5952/changes), I did it like this: `const agent = new ProxyAgent()`, **without** passing a proxy value, and without ternary syntax: `ProxyAgent` lib handle it all automatically.

That’s the [recommended way to initialize the proxy-agent library](https://github.com/TooTallNate/proxy-agents/tree/main/packages/proxy-agent#example), because:

* there are different kinds of proxies (HTTPS, PAC, SOCKS, etc.)
* it automatically respects the `no_proxy` environment variable, so some URLs can bypass the proxy

There are a few other reasons I [mentioned earlier](https://github.com/shadcn-ui/ui/pull/5952):

* it works if `HTTPS_PROXY` (uppercase) is set instead of `https_proxy`
* it supports `http_proxy` (not HTTPS), for cases like using an HTTP `REGISTRY_URL`
* it allows dynamic proxy settings (PAC files)

You also said you tested it manually with `https_proxy=http://<proxy>:<port>`, which is good. But the other cases I mentioned weren’t tested (no offense, I know proxy testing is quite painful to be honest). My proposed code here also handled `no_proxy`, and I verified that behavior in the initial PR.

Comment (nbouvrette) [reply packages/shadcn/src/registry/fetcher.ts]: Thanks @pomeh — you were right that proxy support was thin, and I'd missed this in the first round of the PR.

One small correction on the API: `new ProxyAgent()` (no argument) works for the [`proxy-agent`](https://github.com/TooTallNate/proxy-agents/tree/main/packages/proxy-agent) package you used in #5952, but **undici's `ProxyAgent` is a different class** despite the identical name. From [undici's typings](https://github.com/nodejs/undici/blob/main/types/proxy-agent.d.ts):

```ts
constructor (options: ProxyAgent.Options | string)
```

The `uri` is required — no-arg construction throws.

For the env-var-aware behavior you described, undici ships a separate class: **`EnvHttpProxyAgent`**. I've switched to it in 2784955d9, which now honors:

- `HTTPS_PROXY` / `https_proxy`
- `HTTP_PROXY` / `http_proxy`
- `NO_PROXY` / `no_proxy`

Coverage is locked in by:

- **Unit tests** for the env-detection factory (`packages/shadcn/src/registry/proxy.test.ts`)
- **Integration test** that spins up a real CONNECT-tunneling HTTP proxy and verifies routing + `NO_PROXY` bypass end-to-end (`packages/shadcn/src/registry/proxy.integration.test.ts`)

PAC files and SOCKS proxies aren't supported by `EnvHttpProxyAgent`, but they weren't supported by the previous `https-proxy-agent` setup either — those would require bringing back the broader `proxy-agent` package, which conflicts with this PR's goal of dropping unmaintained transitive deps. Happy to follow up in a separate PR if there's appetite.

Thanks again for catching this.

Comment (nbouvrette) [reply packages/shadcn/src/registry/fetcher.ts]: Update on this thread: I went further than my initial reply scoped — partly because the framing in your original review (this is the natural opportunity to fix the broader proxy gaps, not just the env-var detection) was the right one.

The PR now supports:

| Capability | Before | After |
|---|---|---|
| `HTTPS_PROXY` / `https_proxy` | partial | ✅ |
| `HTTP_PROXY` / `http_proxy` | ❌ | ✅ |
| `NO_PROXY` / `no_proxy` bypass | ❌ | ✅ |
| `ALL_PROXY=socks*://...` (SOCKS4/5) | ❌ | ✅ |
| `PAC_URL=...` (PROXY/SOCKS/DIRECT routing) | ❌ | ✅ |

I was wrong in my earlier reply that adding SOCKS/PAC would require bringing back `proxy-agent` — there's a cleaner path that doesn't reintroduce the abandoned-deps concern:

- **SOCKS** via [`socks`](https://www.npmjs.com/package/socks): plugs directly into undici's dispatcher model via a custom undici `Agent` with a `socks`-based `connect` factory. ~50KB, actively maintained, zero runtime deps.
- **PAC** via [`pac-resolver`](https://www.npmjs.com/package/pac-resolver) + [`quickjs-wasi`](https://www.npmjs.com/package/quickjs-wasi): the JS engine is exposed independently of any HTTP-Agent adapter, so we evaluate PAC and route the result through our own dispatcher selection (HTTP / SOCKS / DIRECT).

PAC support pulls in `quickjs-wasi` (~3 MB WASM blob) for sandboxed PAC script execution. Material for a CLI install but accepted deliberately — PAC scripts can come from network-supplied WPAD discovery where execution isolation matters.

Coverage: **23 unit tests + 13 integration tests** (36 total). Integration tests run a real local CONNECT-tunneling HTTP proxy, a hand-rolled SOCKS5 server, and a local PAC file server — verifying end-to-end routing for each mechanism plus per-host PAC directives.

Title and description on this PR have been updated to reflect the broader scope. Thanks again for the original push.

File: packages/shadcn/src/registry/proxy.integration.test.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+beforeAll(async () => {
+  originServer = createServer((_req, res) => {
+    res.writeHead(200, { "Content-Type": "application/json" })
+    res.end(JSON.stringify({ from: "origin" }))
+  })
+  originUrl = await listen(originServer)
+
+  proxyServer = createServer((req, res) => {
+    proxyHits.push({ url: req.url ?? "", method: req.method ?? "" })
+    const body = JSON.stringify({ from: "proxy-direct" })
```
Comment (pomeh) [inline packages/shadcn/src/registry/proxy.integration.test.ts]: I see 2 strange things here:

1. proxy should not alter the payload from request or response. The origin server responds with `{ from: "origin" }` while proxy server respond with `{ from: "proxy-direct" }`, which are 2 distinct body values. Whether you use a proxy or not, the payload should be the same.
2. I see no other occurrence of value `proxy-direct` elsewhere in this test code. This means the `proxy-direct` is never checked for, which, I guess, can mean several things:
    1. the proxy is never called, thus never return `proxy-direct` JSON value
    2. the proxy is called, but its JSON payload body is totally ignored (and `from: "origin"` would be the only response payload we got in all test cases)
    3. the proxy is called, the `proxy-direct` JSON value is returned, but we never test this value explicitly



To avoid the body alteration problem, and also solve the second point, I would propose:

to not alter the response body payload by the proxy
to add a custom, fake, response header by the proxy, such as `X-E2E-Proxy: e2e-proxy-server`

What make me unsure with this is the fact that `proxyHits` is checked, and it effectively contains the value expected when tunneling through the proxy, meaning the test are working normally. In any case, having `proxy-direct` value here and not elsewhere is somewhat disturbing.

File: packages/shadcn/src/registry/proxy.integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  it("routes through the proxy when destination is not in NO_PROXY list", async () => {
+    process.env.HTTP_PROXY = proxyUrl
+    process.env.NO_PROXY = "example.com"
+    const dispatcher = createProxyDispatcher()
+
+    const response = await fetch(`${originUrl}/test.json`, { dispatcher })
+    const body = (await response.json()) as { from: string }
+    expect(body.from).toBe("origin")
+    expect(proxyHits).toHaveLength(1)
+    expect(proxyHits[0].method).toBe("CONNECT")
```
Comment (pomeh) [inline packages/shadcn/src/registry/proxy.integration.test.ts]: why not perform the exact same test as previously ?

```
      // Proof the proxy was used: a CONNECT was recorded against the origin.
      const originHost = new URL(originUrl).host
      expect(proxyHits).toEqual([{ url: originHost, method: "CONNECT" }])
```

File: packages/shadcn/src/registry/fetcher.ts (hunk 1 above)
Comment (pomeh) [reply packages/shadcn/src/registry/fetcher.ts]: Thanks for these update @nbouvrette , looks like to have full proxy support now, including automated tests, which is nice to ensure things are working as expected. What a great work !

I've done 2 smalls remarks on the updated code (regarding tests).

File: packages/shadcn/src/registry/proxy.integration.test.ts (hunk 2 above)
Comment (nbouvrette) [reply packages/shadcn/src/registry/proxy.integration.test.ts]: You were right that the `proxy-direct` body was dead code — nothing in the test suite asserts on it. The reason is architectural: undici sends `CONNECT host:port` to the proxy (handled by `proxyServer.on("connect", ...)` further down), so the `createServer` request callback is never invoked by any of our actual proxy-routing tests.

Your suggested fix (return the same body as origin + add an `X-E2E-Proxy` header) doesn't quite fit our setup because the proxy uses CONNECT tunneling — it just pipes bytes between client and origin and never sees response headers, so we couldn't add one even if we wanted to.

Instead, in aeaa78b31 I:

1. Replaced the unreachable `proxy-direct` JSON with a clean `502 Bad Gateway` and a comment explaining the proxy is CONNECT-only.
2. Added a TDD-style test that explicitly asserts the 502 behavior, so future edits can't silently reintroduce a code path that no test exercises:

```ts
describe("test proxy server contract", () => {
  it("responds 502 to direct (non-CONNECT) HTTP requests", async () => {
    const response = await fetch(`${proxyUrl}/anything`)
    expect(response.status).toBe(502)
  })
})
```

This makes the test fixture's contract explicit and verifies it. Thanks for the catch.

File: packages/shadcn/src/registry/proxy.integration.test.ts (hunk 3 above)
Comment (nbouvrette) [reply packages/shadcn/src/registry/proxy.integration.test.ts]: Good catch — the inconsistency was unintentional. Tightened in aeaa78b31 to match the parallel test:

```ts
const originHost = new URL(originUrl).host
expect(proxyHits).toEqual([{ url: originHost, method: "CONNECT" }])
```

The looser version (`.toHaveLength(1)` + `.toBe("CONNECT")`) would have missed bugs where the dispatcher CONNECTed to the wrong host (e.g., the proxy's own address instead of the origin's). The strict version catches that bug class.

File: packages/shadcn/src/registry/proxy.integration.test.ts (hunk 2 above)
Comment (pomeh) [reply packages/shadcn/src/registry/proxy.integration.test.ts]: Perfect, this looks good to me now 😃

File: packages/shadcn/src/registry/proxy.integration.test.ts (hunk 3 above)
Comment (pomeh) [reply packages/shadcn/src/registry/proxy.integration.test.ts]: Perfect, this all looks good to me now 💯🚀

**Comment by shadcn** [comment]
I've added this to the roadmap. I'll start looking into this now. Thanks for the great work @nbouvrette This is already looking good.

**Comment by nbouvrette** [comment]
@shadcn merged `main` in and resolved the conflicts — this is green again.

Since #10905 landed the `undici` swap and the `HTTPS_PROXY`/`HTTP_PROXY`/`NO_PROXY` handling from the original version of this PR, I've **rescoped this branch to just the remaining delta: SOCKS and PAC.** Net diff is now +815/−33 across 7 files, down from the original ~854 lines touching the whole fetch stack. Title and description are updated to match.

On the conflict resolution, since `proxy.ts` came back as an add/add:

- **`fetchWithProxy`, `fetchOnce` and `getFailureReason` are preserved verbatim**, including the cross-origin redirect hardening from #11196 and its 5 tests. I took your `fetcher.ts` wholesale.
- The **only** line removed from your version is the module-level `EnvHttpProxyAgent` const, replaced by `createProxyDispatcher(env)`. That factory gates HTTP/HTTPS on the identical four env vars and returns the identical `EnvHttpProxyAgent`, so it's a strict superset — it just checks PAC and SOCKS ahead of it. `git diff` against `main` confirms those 10 lines are the whole removal.
- `proxy.test.ts` is the union of both suites; they cover different exports (`createProxyDispatcher` vs `fetchWithProxy`) with no overlap.
- Lockfile was regenerated from `main`'s so the diff contains only `socks`, `pac-resolver`, `quickjs-wasi` and their transitives.

`pnpm typecheck` is clean and all 1720 tests across 83 files pass.

Two things worth a look when you review:

1. **`quickjs-wasi` is ~3 MB on disk.** It's there so PAC scripts — arbitrary JavaScript fetched over the network — run in a WASM sandbox rather than the CLI's own runtime. If you'd rather not take that weight, I'm happy to split PAC into a separate PR and land SOCKS on its own; SOCKS only needs `socks` (~50 KB) and covers the more common case.
2. **`ALL_PROXY` only activates the SOCKS branch for `socks*://` schemes**, per the curl convention. `ALL_PROXY=http://…` is deliberately left to the explicit `HTTP_PROXY` variable rather than being silently absorbed — happy to change that if you'd prefer it to be permissive.

@axieum the `NO_PROXY` support you asked about shipped in #10905, so that should be working for you on latest already.

@pomeh thanks again for the review earlier — the integration tests you pushed for are still in here and now cover the SOCKS and PAC paths against real local servers.

**Comment by shadcn** [comment]
@nbouvrette Good call on splitting it up. Let's rescope this PR to SOCKS only and move PAC to a follow-up. And keep ALL_PROXY as socks-only. Thanks for working on this. Appreciate it.

(Let's also bump changeset to minor while we're at it).

**Comment by nbouvrette** [comment]
Done — rescoped to SOCKS only. PAC is out of this branch entirely and will come as a follow-up PR once this lands; the work is parked on a branch so nothing's lost.

- **PAC removed**: `PacDispatcher`, the directive parser, `PAC_URL` handling and the `pac-resolver` type stub are gone, along with the `pac-resolver` and `quickjs-wasi` dependencies. That's the ~3 MB of WASM off the dependency tree — `socks` (~50 KB) is now the only addition.
- **`ALL_PROXY` stays SOCKS-only**, as you asked — a non-`socks*` scheme falls through to the explicit `HTTP_PROXY`/`HTTPS_PROXY` variables rather than being silently absorbed. There's a unit test and an integration test pinning that behavior.
- **Changeset bumped to `minor`.**

Net diff is now **+485/−32 across 6 files**, down from +815/−33 across 7. The lockfile diff is 32 lines and is entirely socks-driven: `socks` was already in there as a transitive of `socks-proxy-agent` at 2.8.6, so promoting it to a direct dep moves it to 2.8.9 and updates its `ip-address` dep. Regenerated with the repo's pinned pnpm 10.33.4 and verified with `--frozen-lockfile`.

`fetchWithProxy`/`fetchOnce`/`getFailureReason` are byte-identical to `main` apart from one explanatory comment, so the cross-origin redirect hardening from #11196 is untouched. `pnpm typecheck` is clean and all 1711 tests across 83 files pass.

Title and description are updated to match. The only red check is the Vercel preview, which needs a shadcn-pro member to authorize the deploy from my fork.

**Comment by nbouvrette** [comment]
@shadcn tagging you just in case - let me know if anything needs adjusting! 🙏

**Comment by shadcn** [comment]
Rescope looks great. Thanks @nbouvrette 

Once CI is green. I'll merge this.
